### DISCUSSION TITLE
Console output from loop is printed inline with plot

### DISCUSSION THREAD

**Discussion by AlbertoFabbri93** [quarto-dev/quarto-cli#11032] [Q&A] [answered]
### Description

Hi everybody,
I have the following code:
````qmd
## Gene Set Enrichment Analysis

```{r}
# Set identity to tumor clusters
Idents(breast_cancer_patients_tumor_cells) <- tumor_cluster_name
# Set default assay to RNA
DefaultAssay(breast_cancer_patients_tumor_cells) <- "RNA"

# Run GSEA for each tumor cluster
for (cluster in levels(breast_cancer_patients_tumor_cells)) {
  tumor_cluster_number <- cluster
  print(paste("GSEA analysis of cluster:", tumor_cluster_number))

  # Get the markers for the tumor cluster
  tumor_cluster_markers <- tumor_markers_data %>%
    dplyr::filter(cluster == tumor_cluster_number)
  
  # Count the number of up-regulated genes
  num_upreg_genes <- tumor_cluster_markers %>%
    dplyr::filter(avg_log2FC > 0, p_val < 0.05) %>%
    nrow()
  # Count the number of down-regulated genes
  num_downreg_genes <- tumor_cluster_markers %>%
    dplyr::filter(avg_log2FC < 0, p_val < 0.05) %>%
    nrow()
  
  # How to create the gene list
  # http://yulab-smu.top/biomedical-knowledge-mining-book/faq.html#genelist
  # Gene list creation
  gsea_gene_list <- tumor_cluster_markers %>%
    dplyr::mutate(rank = rank(avg_log2FC, ties.method = "random")) %>%
    dplyr::arrange(desc(rank)) %>%
    with(setNames(avg_log2FC, gene))
  
  # Run GSEA
  gsea <- gseGO(geneList     = gsea_gene_list,
                    keyType      = "SYMBOL",
                    OrgDb        = org.Hs.eg.db, 
                    ont          = "BP",
                    minGSSize    = 10,
                    maxGSSize    = 500,
                    pvalueCutoff = 0.10,
                    verbose      = TRUE,
                    pAdjustMethod = "fdr")
  # Simplify GSEA results by grouping together or eliminating terms that are very similar or redundant
  gsea_simplified <- simplify(gsea, cutoff = 0.7)

  # Do not plot if no pathways have been found
  if (nrow(gsea_simplified@result) > 0) {
    # Prepare plot labels
    GSEAlabels <- labs(
      title = paste("Significant Pathways in Tumor Cluster", tumor_cluster_number), 
      subtitle = "by Gene Set Enrichment Analysis",
      caption = paste0("Based on ", num_upreg_genes," up-regulated genes and ", num_downreg_genes," down-regulated genes."))
    # Prepare plot aesthetics
    GSEAtheme <- theme(
      plot.title = element_text(hjust = 0.5, size = 20),
      plot.subtitle = element_text(hjust = 0.5, size = 15),
      axis.text.x = element_text(angle = 0, size = 11),
      axis.text.y = element_text(angle = 0, size = 8)) 
    # Make the GSEA plots
    GSEADotPlot <- dotplot(gsea_simplified, showCategory = 25) +
      GSEAlabels + labs(x = "Gene Ratio", y = "Biological Pathway") +
      GSEAtheme
    print(GSEADotPlot)
  } else {
    print(paste("No significant pathways found for cluster", tumor_cluster_number))
  }
}
```
````

Inside the for loop I am calling a function that analyze some data and print some information about the process to the console (`verbose = TRUE`). The problem is that the output from this function is only printed to the console for the first cluster, for the subsequent ones the console output is printed above the first plot.

Some images to show you what I mean

The console output with only the info about cluster 0:
![Screenshot 2024-10-11 at 16 01 01](https://github.com/user-attachments/assets/39ba181b-846c-4c88-8624-a83f9ef13dce)

The first plot has the verbose output for all the other runs of geoGO printed above it:
![Screenshot 2024-10-11 at 16 03 49](https://github.com/user-attachments/assets/3a24dd5e-edb9-4251-bede-a22601ef1dea)

The other plots are normal. I tried to create a reproducible example but I am not sure on how to do that. Has anyone an idea on what is going on and how I can get all the verbose output to the console and not together with the first plot?
I tried to fix the problem with `capture.output` and `sink` but they are not able to capture the gseGO verbose output.

Thank you!

EDIT: The verbose output is printed above the first plot even when `verbose = FALSE`

**Comment by mcanouil** [comment]
You can try to make a code that output `message()`, `warning()`, and figures in a loop.

Without reproducible code, there is not much that can be done.

Note that this seems to be an issue with your settings in the context of `knitr` / R, rather than with Quarto CLI itself.

**Reply by AlbertoFabbri93** [reply]
Thank you! Here you have a reproducible example:

````quarto
```{r}
# Loop through 3 iterations
for (i in 1:3) {
  
  # Output a message
  message(paste("Iteration", i, "- This is a message."))
  
  # Generate a plot
  plot(1:10, (1:10) * i, type = 'b', col = i, main = paste("Plot for iteration", i),
       xlab = "X-axis", ylab = "Y-axis")
}
```
````
Result
R Console:
![Screenshot 2024-10-15 at 12 03 23](https://github.com/user-attachments/assets/04ca97cc-51c1-40cd-8def-3891474991fb)
First plot:
![Screenshot 2024-10-15 at 12 03 43](https://github.com/user-attachments/assets/d582ed28-fd42-45cb-bb47-5e451e92ca24)
The other two plots are fine.

Do you have any idea on which settings may be responsible for this behavior?

**Reply by mcanouil** [reply]
Your example is not fully reproducible as it is only a code cell.
Your screenshots seems to be IDE panel window screenshots. If that's the case, Quarto is not responsible of the IDE.

---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

If you have multiple files (and if it is **absolutely required** to have multiple files), please share as a Git repository.

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td></tr>
</table>

Additionally and if not already given, please share the output of `quarto check` within a code blocks (*i.e.*, using three backticks ` ```txt `), see <https://quarto.org/bug-reports.html#check>.

**Reply by AlbertoFabbri93** [reply]
Sorry, this is the full Quarto file:
````qmd
---
title: "Repro messages print together with plot"
format: html
---

```{r}
# Loop through 3 iterations
for (i in 1:3) {
  
  # Output a message
  message(paste("Iteration", i, "- This is a message."))
  
  # Generate a plot
  plot(1:10, (1:10) * i, type = 'b', col = i, main = paste("Plot for iteration", i),
       xlab = "X-axis", ylab = "Y-axis")
}
```
````
Yes, the screenshots are from RStudio. Is this something I should report to them?

quarto check:
```bash
Quarto 1.5.57
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.2.0: OK
      Dart Sass version 1.70.0: OK
      Deno version 1.41.0: OK
      Typst version 0.11.0: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.5.57
      Path: /Applications/RStudio.app/Contents/Resources/app/quarto/bin

[✓] Checking tools....................OK
      TinyTeX: v2024.07.03
      Chromium: (not installed)

[✓] Checking LaTeX....................OK
      Using: TinyTex
      Path: /Users/albi/Library/TinyTeX/bin/universal-darwin
      Version: 2024

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.9.6
      Path: /Library/Developer/CommandLineTools/usr/bin/python3
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with python3 -m pip install jupyter

[✓] Checking R installation...........OK
      Version: 4.4.1
      Path: /Library/Frameworks/R.framework/Resources
      LibPaths:
        - /Users/albi/Documents/University/Research Project/brain-metastases-spatial-omics/renv/library/macos/R-4.4/aarch64-apple-darwin20
        - /Users/albi/Library/Caches/org.R-project.R/R/renv/sandbox/macos/R-4.4/aarch64-apple-darwin20/f7156815
      knitr: 1.48
      rmarkdown: 2.28

[✓] Checking Knitr engine render......OK
```

**Reply by cderv** [reply]
This is expected that message is printed inside a cell output with **knitr** inside Quarto considering default behavior. 

You need to set `message: false` is you want message to be show in console output only. 
````markdown
```{r}
#| message: false
# Loop through 3 iterations
for (i in 1:3) {
  
  # Output a message
  message(paste("Iteration", i, "- This is a message."))
  
  # Generate a plot
  plot(1:10, (1:10) * i, type = 'b', col = i, main = paste("Plot for iteration", i),
       xlab = "X-axis", ylab = "Y-axis")
}
```
````

This is a knitr behavior. You can control message, warning and error output for the code computation. 

Hope it helps

**Reply by AlbertoFabbri93** [reply]
@cderv Thank you for your answer, but if I set `message: false` or `warning: false` then I have no way to see these messages/warnings. In RStudio they do not show up in the R console output any longer.

**Reply by mcanouil** [reply]
You need to use the R/knitr way of setting the option in a code cell.

```r
knitr::opts_chunk$set(message = NA)
```

NA is not a cross-engine valid value in Quarto.

See <https://yihui.org/knitr/options/>.

**Reply by cderv** [reply]
> In RStudio they do not show up in the R console output any longer.

@AlbertoFabbri93 sorry I missed this requirement. Yeah you need some tweak for them to show in logged output. 

Let me complement @mcanouil answer to explain what is happening here. 

There was a required changed in 1.42 due to **evaluate** R package change that you can find in here: https://github.com/yihui/knitr/blob/master/NEWS.md#major-changes-4

> Due to a change in the evaluate package, the chunk options message = FALSE and warning = FALSE will completely suppress the messages/warnings now, instead of sending them to the console. To get back to the old behavior, you can use NA instead of FALSE 

So if you want message and warning to go in console instead of document, you need to set to `NA` the R option. 

In Quarto this is different that in R Markdown due to how options are set using YAML only. So here are the solution. 

- A setup chunk in your .qmd has mentionned above with `knitr::opts_chunk$set(message = NA)` => that is the easiest way.
- You can do per chunk configuration, only valid for **knitr**, using 
	````
	#| message: !expr NA
	````
	`!expr` is a special syntax for **knitr** engine so it can be used only inside chunk. 
- If you want to do it globally, for all document you can also, or on your machine, you can configure in .Rprofile (global or project one) using `options(knitr.chunk.message = NA)`
- A hidden hack is using environment variable too if that is easier : 
	```
	R_KNITR_OPTIONS="knitr.chunk.message = NA, knitr.chunk.warning = NA"
	```
	I think it will work. 

Anyhow, I'll open an issue in **knitr** so that we see how we can change NA only config to something that is supported by YAML spec directly (NA is not)

Thanks for the feedback.

**Reply by mcanouil** [reply]
Christophe, what is the precedence of options here? Environment < YAML < R code ?

**Reply by cderv** [reply]
Same as in **knitr**. Closest one to cell wins. 

Like using `knitr::opts_chunk$set(message = NA)`, if calls twice at different places, last one wins. Setting R options does end up setting `knitr::opts_chunk$set`. And if a cell as `message: true` then, this last configuration will win. 

I believe this is quite common behavior with configuration, where usually latest set wins.

**Reply by mcanouil** [reply]
Indeed, I wanted to confirm that was the case here as well.

**Reply by AlbertoFabbri93** [reply]
> In Quarto this is different that in R Markdown due to how options are set using YAML only. So here are the solution.
> 
>  * A setup chunk in your .qmd has mentionned above with `knitr::opts_chunk$set(message = NA)` => that is the easiest way.
> 
>  * You can do per chunk configuration, only valid for **knitr**, using
>    ```
>     #| message: !expr NA
>    ```

I do not know if I am missing something here but I really can not get this to work. I have tried the first option by adding the code to a chunk above the loop and nothing changes. Same thing with adding the cell options to the top of the chunk with the loop:

![Screenshot 2024-10-18 at 13 09 52](https://github.com/user-attachments/assets/75b9a73c-b5bd-4b17-8b57-a954b6237a6d)

**Reply by mcanouil** [reply]
Are you expecting this to work in the editor view?
Your screenshot is neither the rendered document, nor the terminal.

Try rendering your document independently of the IDE with `quarto render your-file.qmd`.

I believe your issue is unrelated to Quarto (or Rmarkdown).

**Reply by cderv** [reply, marked as answer]
RStudio Notebook view with inline output is not running **knitr**. So this a Rstudio issue on how they handle message and warning. 

I missed that from your initial message - you should search rstudio IDE issues, and ask there. They probably do not support `message: NA` either or maybe it is something else. 

Anything in notebook view is IDE handling the cells, not **knitr**.

**Reply by AlbertoFabbri93** [reply]
Ah ok, makes sense. Thank you very very much for all the explanations, I learned a lot!

**Reply by cderv** [reply]
> You need to use the R/knitr way of setting the option in a code cell.
> 
> ```r
> knitr::opts_chunk$set(message = NA)
> ```
> 
> NA is not a cross-engine valid value in Quarto.
> 
> See [yihui.org/knitr/options](https://yihui.org/knitr/options/).

With dev version of **knitr**, you can now set `#| message: NA` is a code cell. 
https://github.com/yihui/knitr/issues/2375

This will we converted from `"NA"` to `NA` by **knitr** and have the right effect. No need to change globally using the knit options. 

I'll update YAML validation for this

**Reply by mcanouil** [reply]
That's awesome. (I can get rid of `knitr::opts_chunk`)

**Reply by cderv** [reply]
> (I can get rid of knitr::opts_chunk)

Yes that this is the purpose 😄

**Reply by mcanouil** [reply]
~~Could you do the same for `warning` which also supports `NA` to print them to console?~~

Nevermind.

>  If a character value is passed to the chunk option `message` or `warning`, it will be coerced by `as.logical()`, e.g., a character string `"NA"` will be coerced to `NA` (thanks, @cderv, #2375).
